src/charm.py

```diff
@@ -0,0 +1,298 @@
+    ) -> None:
+        """Handle db-relation-joined.
+
+        Here we set the database name for the synapse charm and
+        extensions needed.
+
+        Args:
+            event: Event triggering the database relation joined handler.
+        """
+        if self.model.unit.is_leader():
+            # Provide requirements to the PostgreSQL server.
+            event.database = DATABASE_NAME
+            event.extensions = ["hstore:public", "pg_trgm:public"]
+        elif event.database != DATABASE_NAME:
+            # Leader has not yet set requirements. Defer, in case this unit
+            # becomes leader and needs to perform that operation.
+            event.defer()
+
+    # pgsql.MasterChangedEvent is actually defined
+    def _on_master_changed(self, event: pgsql.MasterChangedEvent) -> None:
+        """Handle changes in the primary database unit.
+
+        Args:
+            event: Event triggering the database master changed handler.
+        """
+        if event.master is None:
+            self._stored.db_name = None
+            self._stored.db_user = None
+            self._stored.db_password = None
+            self._stored.db_host = None
+            self._stored.db_port = None
+        else:
+            self._stored.db_name = event.master.dbname
+            self._stored.db_user = event.master.user
+            self._stored.db_password = event.master.password
+            self._stored.db_host = event.master.host
+            self._stored.db_port = event.master.port
+
+        self._on_config_changed(event)
+
+    def _on_start(self, _):
+        """Generate a config template to be rendered later.
+
+        Generating with environment variables will render
+        a config file with PostgreSQL enabled for instance
+        instead of the default SQLite backend."""
+        self._run_generate_synapse()
+
+    @property
+    def _external_hostname(self):
+        """Return the external hostname passed to ingress via the relation."""
+        # It is recommended to default to `self.app.name` so that the external
+        # hostname will correspond to the deployed application name in the
+        # model, but allow it to be set to something specific via config.
+        return self.config["server_name"] or self.app.name
+
+    def _container(self):
+        """Get the Synapse workload container.
+
+        Returns:
+            The pebble instance of the Synapse container.
+        """
+        return self.unit.get_container(self._CONTAINER_NAME)
+
+    def _populate_synapse_env_settings(self):
+        """Populate env settings to pass around."""
+        if os.getenv("SYNAPSE_SERVER_NAME") is None:
+            os.environ["SYNAPSE_SERVER_NAME"] = self.config["server_name"]
+        if os.getenv("SYNAPSE_REPORT_STATS") is None:
+            os.environ["SYNAPSE_REPORT_STATS"] = self.config["report_stats"]
+        pod_config = {
+                "SYNAPSE_SERVER_NAME": os.getenv("SYNAPSE_SERVER_NAME"),
```

Comment:
I am a little curious why not directly retrieve the server name from configuration, instead of saving the variable in the charm environment variables then put into this env setting dict?

```python
        pod_config = {
                 "SYNAPSE_SERVER_NAME": self.config["server_name"],
        ...
        }
```

Reply:
Well (for now) the server name has a default of "synapse-local" which should be a default fallback in case you want to spin up synapse locally (and have synapse-local resolve to 127.0.0.1). The docker version of synapse relies on you exposing these two env variables before starting the service and here I'm replicating that behavior - if you haven't set these variables, you would use the config fallback. That is not the best thing, changing the server name very much creates a new synapse server under a new name, so some thought needs to be put as to how to manage that.

Reply:
The charm environment variables is not persisted across events. For example, when charm finished handling the `config-changed` event, charm will actual terminate itself. And for the next event, for example, `pebble-ready`, a new charm process will spawn and the charm environment variables set during the last event handling will be gone. Is this acceptable?

Reply:
I think I get your point - if I understand correctly, that way I will always set the SYNAPSE_SERVER_NAME to self.config["server_name"], as on any new event, it does not get populated (and does not persist from a previous event). Then the proper way perhaps is to have a default on "server_name" to be "" and a check in the src/charm.py that blocks the charm if it is "", so until we set it to something that we intend to use. Thoughts?

Reply:
See my comment in the config.yaml concerning this option